Implement bridge

xepozz opened the pull request:
| Q             | A
| ------------- | ---
| Is bugfix?    | ✔️/❌
| New feature?  | ✔️/❌
| Breaks BC?    | ✔️/❌
| Fixed issues  | comma-separated list of tickets # fixed by the PR, if any

xepozz committed:
Implement sentry bridge

samdark committed:
Apply fixes from StyleCI

samdark commented on composer.json:
```diff
@@ -19,7 +19,10 @@
     "require": {
-        "php": "^7.4|^8.0"
+        "php": "^7.4|^8.0",
+        "guzzlehttp/guzzle": "^7.3",
```
I'm not sure it is a good idea to enforce usage of Guzzle. How about using PSR-18 instead?

xepozz replied:
Agreed, but
```
$ composer req sentry/sentry
Using version ^3.3 for sentry/sentry
./composer.json has been updated
Running composer update sentry/sentry
Loading composer repositories with package information
Updating dependencies
Your requirements could not be resolved to an installable set of packages.

  Problem 1
    - sentry/sentry[dev-master, 3.3.0] require php-http/async-client-implementation ^1.0 -> could not be found in any version, but the following packages provide it:
      - symfony/http-client Provides powerful methods to fetch HTTP resources synchronously or asynchronously
      - php-http/guzzle6-adapter Guzzle 6 HTTP Adapter
      - symfony/symfony The Symfony PHP framework
      - php-http/curl-client PSR-18 and HTTPlug Async client with cURL
      - php-http/guzzle7-adapter Guzzle 7 HTTP Adapter
      - php-http/mock-client Mock HTTP client
      - voku/httpful A Readable, Chainable, REST friendly, PHP HTTP Client
      - php-http/react-adapter React HTTP Adapter
      - swisnl/php-http-fixture-client Fixture client for PHP-HTTP
      - phpgt/fetch Asynchronous HTTP client with promises for PHP 7 applications.
      - betalabs/guzzle6-adapter Guzzle 6 HTTP Adapter
      - neur0toxine/pock PSR-18 compatible HTTP mock library
      - nessworthy/amphp-http-httplug A small crappy adapter package to support php-http/async-client-implementation via. amphp/http-clien
      - phpgt/curl cURL object wrapper.
      - john_zuk/php-nbp-api NBP (Narodowy Bank Polski) API client
      - christophwurst/nextcloud-http-client HTTPlug and PSR7 adapter for Nextcloud
      - devster/http-client Symfony HttpClient component
      - gogoprint/guzzle6-adapter Guzzle 6 HTTP Adapter
      - pbxapi/php-http-fixture-client Fixture client for PHP-HTTP
      - phpgt/curlinterface cURL object wrapper.
      ... and 2 more.
      Consider requiring one of these to satisfy the php-http/async-client-implementation requirement.
    - sentry/sentry 3.3.x-dev is an alias of sentry/sentry dev-master and thus requires it to be installed too.
    - Root composer.json requires sentry/sentry ^3.3 -> satisfiable by sentry/sentry[3.3.0, 3.3.x-dev (alias of dev-master)].

Potential causes:
 - A typo in the package name
 - The package is not available in a stable-enough version according to your minimum-stability setting
   see <https://getcomposer.org/doc/04-schema.md#minimum-stability> for more details.
 - It's a private package and you forgot to add a custom repository to find it

Read <https://getcomposer.org/doc/articles/troubleshooting.md> for further common problems.

Installation failed, reverting ./composer.json and ./composer.lock to their original content.
```

samdark replied:
Well, you can add Guzzle as a dev dependency, but not as production one.

xepozz replied:
Done

vjik commented on composer.json:
```diff
@@ -40,6 +43,10 @@
         "branch-alias": {
             "dev-master": "3.0.x-dev"
+        },
+        "config-plugin": {
+            "params": "config/params.php",
+            "common": "config/common.php"
```
```suggestion
        "config-plugin-options": {
            "source-directory": "config" 
        },
        "config-plugin": {
            "params": "params.php",
            "common": "common.php"
```

rustamwin commented on config/providers.php:
```diff
@@ -0,0 +1,9 @@
+
+use Yiisoft\Yii\Sentry\Provider\SentryProvider;
+
+return [
```
What about ability to disable provider in params?

xepozz committed:
Add middleware

xepozz committed:
Use hub instance from container

xepozz committed:
Remove gitkeep

xepozz committed:
Adjust composer.json

xepozz committed:
Add hub definition

samdark committed:
Apply fixes from StyleCI

xepozz committed:
Add ability to disable sentry

xepozz committed:
Merge remote-tracking branch 'origin/implement-bridge' into implement-bridge

samdark commented on src/Middleware/SentryIntegration.php:
```diff
@@ -0,0 +1,32 @@
+use Sentry\State\HubInterface;
+use Throwable;
+
+final class SentryIntegration implements MiddlewareInterface
```
MiddlewareInterface requires `psr/http-server-middleware` in composer's `require`.

xepozz replied:
Done

xepozz committed:
Add missing dependencies and sort exists

samdark commented:
@xepozz is it still work in progress or ready for review?

xepozz commented:
It's ready. Another features will be added later.

samdark commented on composer.json:
```diff
@@ -19,9 +19,16 @@
     "require": {
-        "php": "^7.4|^8.0"
+        "php": "^7.4|^8.0",
+        "php-http/httplug": "^2.2",
```
I think it worth moving this dependency to require-dev and requiring https://packagist.org/providers/psr/http-client-implementation instead.

xepozz replied:
It's impossible. Sentry needs async http client. 
See https://github.com/yiisoft/yii-sentry/pull/1#discussion_r650594020

samdark replied:
Can't we depend on [HTTP client PSR](https://www.php-fig.org/psr/psr-18/)? `psr/http-client-implementation`

xepozz replied:
No

samdark commented on composer.json:
```diff
@@ -19,9 +19,16 @@
+        "yiisoft/di": "3.0.x-dev"
     },
     "require-dev": {
+        "guzzlehttp/guzzle": "^7.3",
```
Why getting both guzzle and http plug at the same time?

xepozz replied:
httpplug needs to configure abstractions in config files. [httplug doesn't provide an adapter](https://github.com/php-http/httplug/tree/master/src).

And yes, I need to move `guzzle` to `require` section too. Are you agree with it or not?

samdark replied:
I'd prefer not to tie to concrete implementation.

samdark commented on src/Middleware/SentryIntegration.php:
```diff
@@ -0,0 +1,32 @@
+use Sentry\State\HubInterface;
+use Throwable;
+
+final class SentryIntegration implements MiddlewareInterface
```
```suggestion
final class SentryMiddleware implements MiddlewareInterface
```

samdark commented:
Since there are only two classes, I think we can put them to the package root namespace without sub-namespaces.

xepozz committed:
Rename and move classes

vjik commented on src/SentryProvider.php:
```diff
@@ -0,0 +1,37 @@
+
+final class SentryProvider extends ServiceProvider
+{
+    /**
+     * @psalm-suppress InaccessibleMethod
+     */
```
```suggestion
```

no need

xepozz committed:
Update src/SentryProvider.php

Co-authored-by: Sergei Predvoditelev <sergey.predvoditelev@gmail.com>

samdark commented:
@xepozz let's require defining your own PSR implementation and use interface instead.

xepozz committed:
Move http driver to application template

xepozz committed:
Add http driver to dev requirements

xepozz commented:
I moved packages to dev requirements and moved configuration into userland.
See https://github.com/yiisoft/app-api/pull/85 if you want to try how it works.

samdark commented on config/common.php:
```diff
@@ -0,0 +1,22 @@
+    \Sentry\HttpClient\HttpClientFactoryInterface::class => function (Yiisoft\Injector\Injector $injector) {
+        return $injector->make(\Sentry\HttpClient\HttpClientFactory::class, [
+            'sdkIdentifier' => \Sentry\Client::SDK_IDENTIFIER,
+            // TODO use composer tool
```
?

xepozz replied:
Fixed

samdark commented on config/common.php:
```diff
@@ -0,0 +1,22 @@
+return [
+    \Sentry\Transport\TransportFactoryInterface::class => \Sentry\Transport\DefaultTransportFactory::class,
+    \Sentry\HttpClient\HttpClientFactoryInterface::class => function (Yiisoft\Injector\Injector $injector) {
+        return $injector->make(\Sentry\HttpClient\HttpClientFactory::class, [
```
I think we can safely do `new` here instead of using injector.

xepozz replied:
Not really, but I rewrote syntax to array-like

samdark commented on config/common.php:
```diff
@@ -0,0 +1,22 @@
+
+return [
+    \Sentry\Transport\TransportFactoryInterface::class => \Sentry\Transport\DefaultTransportFactory::class,
+    \Sentry\HttpClient\HttpClientFactoryInterface::class => function (Yiisoft\Injector\Injector $injector) {
```
```suggestion
    \Sentry\HttpClient\HttpClientFactoryInterface::class => static function (Yiisoft\Injector\Injector $injector) {
```

xepozz committed:
Update config/common.php

Co-authored-by: Alexander Makarov <sam@rmcreative.ru>

xepozz committed:
Rewrite config with array-like syntax

xepozz committed:
Use composer tool for detecting package version
